# Genotyping chrX

To genotype based on depth (remember to change number of zeros depending on number of samples):
```
./10_Genotypes_only_male_chrX_based_on_allelic_depth.pl /work/ben/2017_rhesus_genomez/F_and_M/FandM_chrX_BSQR_jointgeno_allsites_filtered.vcf.gz 0000000000000000000000000000000000000000 /work/ben/2017_rhesus_genomez/F_and_M/FandM_chrX_allhaploid_BSQR_jointgeno_allsites_filtered.vcf.gz.tab
```

I still need to modify this script to match the hardfilter criteria that are applied to the aDNA data.  For example, depth of coverage, etc.
