# Genotyping chrX

To genotype based on depth (remember to change number of zeros depending on number of samples):
```
./10_Genotypes_only_male_chrX_based_on_allelic_depth.pl /work/ben/2017_rhesus_genomez/F_and_M/FandM_chrX_BSQR_jointgeno_allsites_filtered.vcf.gz 0000000000000000000000000000000000000000 /work/ben/2017_rhesus_genomez/F_and_M/FandM_chrX_allhaploid_BSQR_jointgeno_allsites_filtered.vcf.gz.tab
```

I still need to modify this script to match the hardfilter criteria that are applied to the aDNA data.  For example, depth of coverage, etc.

Another (BAD) option is to use bcftools to fix ploidy after they have already gotten diploid genotypes, like this:
```
zcat FandM_chrX_BSQR_jointgeno_allsites_filtered.vcf.gz | /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/bin/bcftools-1.6/bcftools +fixploidy -O z -o /scratch/ben/FandM_chrX_BSQR_jointgeno_allsites_filtered_haploid.vcf.gz -- -s /scratch/ben/male_samplez 
```
this is BAD because bcftools outputs the first (REF) allele irrespective of depth.  Better to use my script.

I need to modify my script to output a vcf file.

I made a sample to work with like this:
```
gunzip -c FandM_chrX_BSQR_jointgeno_allsites_withpapio_filtered2_coverage_SNPsonly.vcf.gz.recode.vcf.gz | sed -ne '1,200p' > temnp.vcf
```

