# Genotyping and filtering chrX 
In this directory (`/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/ben_scripts`, I have these sbatch files:
```sh
2020_chrX_0_gatk_HaplotypeCaller_chrX_males.sh		  
2020_chrX_1_gatk_GenotypeGVCFs_chrX_females_and_males.sh  
2020_chrX_2_gatk_SelectVariants_indels_chrX.sh
2020_chrX_3_gatk_VariantFiltration_chrX.sh
2020_chrX_4_gatk_SelectVariants_chrX.sh
2020_chrX_5_vcftools_delete_PAR.sh
2020_chrX_6_gatk_SelectVariants_SNPsonly_chrX.sh
2020_chrX_7_gatk_VariantsToTable_chrX.sh
```
The first one was used to re-call genotypes in males with ploidy equal to 1:
```
#!/bin/sh
#SBATCH --job-name=haplotypecaller
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=24:00:00
#SBATCH --mem=32gb
#SBATCH --output=haplotypecaller.%J.out
#SBATCH --error=haplotypecaller.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T HaplotypeCaller -R ../../MacaM/MacaM_mt_female.fa -I /home/ben/projects/r
rg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/males/${1} -ploidy 1 -L chrX --emitRefConfidence GVCF -o ${2}_hap
loid_chrX_BSQR.g.vcf.gz
```
the second one selects indels for filtering:
```
#!/bin/sh
#SBATCH --job-name=indels
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=12:00:00
#SBATCH --mem=32gb
#SBATCH --output=indels.%J.out
#SBATCH --error=indels.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T SelectVariants -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/Ma
caM/MacaM_mt_female.fa -V all_diploid_haploid_chrX_BSQR.vcf.gz -selectType INDEL -o /home/ben/projects/rrg-ben/ben/2017
_SEAsian_macaques/SEAsian_macaques_bam/with_papio/all_diploid_haploid_chrX_BSQR_indels.vcf.gz

```
the third one flags bad positions and genotypes, similar to autosomal filtering except allowing a lower minimum
depth (5 instead of 10) for chrX in males
```
#!/bin/sh
#SBATCH --job-name=VarFilt
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=24:00:00
#SBATCH --mem=32gb
#SBATCH --output=VarFilt.%J.out
#SBATCH --error=VarFilt.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T VariantFiltration -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques
/MacaM/MacaM_mt_female.fa -V all_diploid_haploid_chrX_BSQR.vcf.gz --filterExpression "QD < 2.0 || FS > 30.0 || MQ < 40.
0 || ReadPosRankSum < -8.0 || ExcessHet > 15.00 || MQRankSum < -10.00 || InbreedingCoeff < 0.00" --filterName "lowqual"
 -mask /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio/all_diploid_haploid_chrX_BS
QR_indels.vcf.gz --maskExtension 5 --maskName "indel" --genotypeFilterExpression "DP < 5 || DP > 100" --genotypeFilterN
ame "genotypefilter" --setFilteredGtToNocall -o /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_b
am/with_papio/all_diploid_haploid_chrX_BSQR_flagged1.vcf.gz
```
the fourth one filters these flagged sites:
```
#!/bin/sh
#SBATCH --job-name=selectVar
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=12:00:00
#SBATCH --mem=32gb
#SBATCH --output=selectvar.%J.out
#SBATCH --error=selectvar.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T SelectVariants -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/Ma
caM/MacaM_mt_female.fa -V /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio/all_dipl
oid_haploid_chrX_BSQR_flagged1.vcf.gz --excludeFiltered --setFilteredGtToNocall --maxNOCALLfraction 0.5 -o /home/ben/pr
ojects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio/all_diploid_haploid_chrX_BSQR_filtered1.vcf.gz
```
And the fifth one deletes the PAR:
```
#!/bin/sh
#SBATCH --job-name=vcftools
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=12:00:00
#SBATCH --mem=3gb
#SBATCH --output=vcftools.%J.out
#SBATCH --error=vcftools.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load intel/2018.3
module load vcftools/0.1.16

vcftools --gzvcf /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio/all_diploid_haplo
id_chrX_BSQR_filtered1.vcf.gz --out /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papi
o/all_diploid_haploid_chrX_BSQR_filtered2_noPAR.vcf.gz --chr chrX --from-bp 1691776 --to-bp 148935249 --recode

```
And the sixth one selects SNPs only:
```
#!/bin/sh
#SBATCH --job-name=SNPsonly
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=12:00:00
#SBATCH --mem=32gb
#SBATCH --output=SNPsonly.%J.out
#SBATCH --error=SNPsonly.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T SelectVariants -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/Maca
M/MacaM_mt_female.fa -V /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio/all_diploid_
haploid_chrX_BSQR_filtered2_noPAR.vcf.gz.recode.vcf.gz -selectType SNP -o /home/ben/projects/rrg-ben/ben/2017_SEAsian_mac
aques/SEAsian_macaques_bam/with_papio/all_diploid_haploid_chrX_BSQR_filtered2_noPAR_SNPsonly.vcf.gz.recode.vcf.gz
```

And the seventh one will generate a table that will be used for filtering based on depth (described on a different github page):
```
#!/bin/sh
#SBATCH --job-name=Var_to_table
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=12:00:00
#SBATCH --mem=32gb
#SBATCH --output=Vartotable.%J.out
#SBATCH --error=vartotable.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -Xmx2g  -jar $EBROOTGATK/GenomeAnalysisTK.jar -T VariantsToTable -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_m
acaques/MacaM/MacaM_mt_y.fa -V ../SEAsian_macaques_bam/with_papio/all_diploid_haploid_chrX_BSQR_filtered2_noPAR.vcf.gz
.recode.vcf.gz -F CHROM -F POS -GF DP -o ../SEAsian_macaques_bam/with_papio/GVCF_chrX_gf_DP.table

```




# Anything blow this line ultimately was not used....

To genotype based on depth (remember to change number of zeros depending on number of samples):
```
./10_Genotypes_only_male_chrX_based_on_allelic_depth.pl /work/ben/2017_rhesus_genomez/F_and_M/FandM_chrX_BSQR_jointgeno_allsites_filtered.vcf.gz 0000000000000000000000000000000000000000 /work/ben/2017_rhesus_genomez/F_and_M/FandM_chrX_allhaploid_BSQR_jointgeno_allsites_filtered.vcf.gz.tab
```

I still need to modify this script to match the hardfilter criteria that are applied to the aDNA data.  For example, depth of coverage, etc.

Another (BAD) option is to use bcftools to fix ploidy after they have already gotten diploid genotypes, like this:
```
zcat FandM_chrX_BSQR_jointgeno_allsites_filtered.vcf.gz | /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/bin/bcftools-1.6/bcftools +fixploidy -O z -o /scratch/ben/FandM_chrX_BSQR_jointgeno_allsites_filtered_haploid.vcf.gz -- -s /scratch/ben/male_samplez 
```
this is BAD because bcftools outputs the first (REF) allele irrespective of depth.  Better to use my script.

First I deleted the PAR region, which is pos 1-1691775 in MacaM based on comparison to the rhesus chrY, like this (this command saves the non-PAR region):
```
vcftools --gzvcf FandM_chrX_BSQR_jointgeno_allsites_withpapio_filtered2_coverage_SNPsonly.vcf.gz.recode.vcf.gz --out FandM_chrX_BSQR_jointgeno_allsites_withpapio_filtered3_coverage_SNPsonly.vcf.gz --chr chrX --from-bp 1691776 --to-bp 148935249 --recode
```

I need to modify my script to output a vcf file.

I made a sample to work with like this:
```
gunzip -c FandM_chrX_BSQR_jointgeno_allsites_withpapio_filtered2_coverage_SNPsonly.vcf.gz.recode.vcf.gz | sed -ne '1,200p' > temnp.vcf
```

And I wrote this script (10.1_Genotypes_only_male_chrX_based_on_allelic_depth_outputvcf.pl) to output a new vcf file based on allelic depth only for the males.
```
#!/usr/bin/env perl
use strict;
use warnings;
use List::Util 'max';
use List::Util qw(shuffle);

# This program reads in a vcf file then genotypes chrX sequences
# based on the AD (allelic depth) annotation for males
# Females are left as is

# To run type this:
# 10.1_Genotypes_only_male_chrX_based_on_allelic_depth_outputvcf.pl temnp.vcf 111111111000110000110001111110 output.vcf

# where 0101010 indicates for each ingroup 
# sample whether the individual is not (0) or is (1)
# a female

# It takes as input a vcf file and outputs a tab delimited file


my $inputfile = $ARGV[0];
my $outputfile = $ARGV[2];

unless (open(OUTFILE, ">$outputfile"))  {
	print "I can\'t write to $outputfile   $!\n\n";
	exit;
}
print "Creating output file: $outputfile\n";


if ($inputfile =~ /.gz$/) {
	open(DATAINPUT, "gunzip -c $inputfile |") || die "can’t open pipe to $inputfile";
}
else {
	open(DATAINPUT, $inputfile) || die "can’t open $inputfile";
}


my @sexes = split("",$ARGV[1]);

my $y;
my $x;
my @columns=();
my @fields;
my $AD;
my $GT;
my $DP;
my $counter=0;
my @genotypes;
my @genotypez;
my @otherstuff;
my @alleledepth;
my @alleledepthz;
my @readdepth_DPz;
my $smush;
my $a;
my $max;
my @maxcounter=();
my $counter2=0;
my @altalleles=();
my @allelieos;

while ( my $line = <DATAINPUT>) {
	chomp($line);
	@columns=split("\t",$line);
		if(substr($columns[0],0,1) ne '#'){ # this is not a comment
			@fields=split(":",$columns[8]);
			$counter=0;
			$AD=0;
			$GT=0;
			$DP=0;
			@altalleles=split(",",$columns[4]);
			# first find out where the AD and GT columns are
			foreach(@fields){
				if($_ eq 'AD'){
					$AD=$counter;
					if($AD != 1){
						print "Problem - the AD field was not second\n";
						print $line;
					}
				}
				elsif($_ eq 'GT'){
					$GT=$counter;
					if($GT != 0){
						print "Problem - the GT field was not first\n";
					}
				}
				elsif($_ eq 'DP'){
					$DP=$counter;
					if($DP != 2){
						print "Problem - the DP field was not third\n";
					}
				}
				$counter+=1;
			}
			# print out the first 9 columns
			for ($y = 0 ; $y <= 8; $y++ ) {
				print OUTFILE $columns[$y],"\t";
			}
			print OUTFILE $columns[9];
			# now load genotypes into an array @genotypez
			# and the allele depths into an array @alleledepthz
			# and the other stuff into an array
			@genotypes=();
			@otherstuff=();
			@readdepth_DPz=();
			@alleledepthz=();
			for ($y = 9 ; $y <= $#columns; $y++ ) {
				@genotypes=split(":",$columns[$y]);
				push @genotypez,$genotypes[$GT];
				push @alleledepthz,$genotypes[$AD];
				push @readdepth_DPz,$genotypes[$DP];
				$smush="";
				for ($a = 3 ; $a <= $#genotypes; $a++ ) { # add all the other stuff
					# will be added later
					$smush = $smush.':'.$genotypes[$a];
				}	
				push @otherstuff, $smush;
			}
			#print "hi ",$#genotypes," gu ",@otherstuff,"\n";
			# now we have an array with GT, AD, DP, and the rest 
			for ($y = 9 ; $y <= $#columns; $y++ ) { # go through each individual
				if($sexes[$y-9] == 0){ # this is a male
					@alleledepth=();
					@genotypes=();
					@allelieos=();
					@genotypes=split(":",$columns[$y]);
					@alleledepth=split(",",$genotypes[$AD]);
					@allelieos = split("/",$genotypes[$GT]); # these are the alleles with numbers 0, 1, 2, etc
					@maxcounter=();
					$counter2=0;
					$max=0;
					$max=max @alleledepth;
					# now cycle through each allele depth to find highest and see if there is a tie
					foreach my $alleledepth (@alleledepth){
						if($alleledepth == $max){
							push(@maxcounter,$counter2);
						}
						$counter2+=1;
					}	
					@maxcounter = shuffle @maxcounter; # pick a random allele if there is a tie
					# print the entry for this sample (use max allele depth, with the second one becoming the DP entry)
					if($allelieos[0] eq "."){
						print OUTFILE "\t\.:",$alleledepth[$maxcounter[0]],":",
						$alleledepth[$maxcounter[0]],$otherstuff[$y-9];
					}
					else{
						print OUTFILE "\t",$maxcounter[0],":",$alleledepth[$maxcounter[0]],":",
						$alleledepth[$maxcounter[0]],$otherstuff[$y-9];						
					}	
				}
				else{ # this is a female
					print OUTFILE "\t",$columns[$y];
				}
			} # end y for loop
			print OUTFILE "\n";	# end of the SNP
		}# endif
		else{ # print the comment to the outputfile
			print OUTFILE $line,"\n";;			
		}
}# end while
close DATAINPUT;
close OUTFILE;
```

# 2020 UPDATE to Genotyping chrX
All the approaches above don't work 100% for various reasons.  So I went back to BSQRed bam files and recalled chrX using Haplotype caller with ploidy = 1 for males.
```
#!/bin/sh
#SBATCH --job-name=haplotypecaller
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=24:00:00
#SBATCH --mem=32gb
#SBATCH --output=haplotypecaller.%J.out
#SBATCH --error=haplotypecaller.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T HaplotypeCaller -R ../../MacaM/MacaM_mt_female.fa -I /home/ben/projects/rrg-ben/be
n/2017_SEAsian_macaques/SEAsian_macaques_bam/males/${1} -ploidy 1 -L chrX --emitRefConfidence GVCF -o ${2}_haploid_chrX_BSQR.g.v
cf.gz

```

Now I have each of these chrX files for each sample in directory:
```
/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio
```
This then needs to be converted to a vcf file using GenotypeGVCF:2020_chrX_1_gatk_GenotypeGVCFs_chrX_females_and_males.sh
```
#!/bin/sh
#SBATCH --job-name=GenotypeGVCFs
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=24:00:00
#SBATCH --mem=32gb
#SBATCH --output=GenotypeGVCFs.%J.out
#SBATCH --error=GenotypeGVCFs.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T GenotypeGVCFs -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/Mac
aM/MacaM_mt_female.fa -V bru_PF707_chrX_BSQR.g.vcf.gz -V ../SEAsian_macaques_bam/papio_g_vcfs/Papio_rg_realigned_dedup.
bamBSQR.bam_chrX_BSQR.g.vcf.gz -V hecki_PF505_chrX_BSQR.g.vcf.gz -V hecki_PF643_chrX_BSQR.g.vcf.gz -V hecki_PF644_chrX_
BSQR.g.vcf.gz -V hecki_PF647_chrX_BSQR.g.vcf.gz -V hecki_PF648_chrX_BSQR.g.vcf.gz -V maura_PF615_chrX_BSQR.g.vcf.gz -V 
maura_PF713_chrX_BSQR.g.vcf.gz -V maura_PM613_haploid_chrX_BSQR.g.vcf.gz -V maura_PM614_haploid_chrX_BSQR.g.vcf.gz -V m
aura_PM616_haploid_chrX_BSQR.g.vcf.gz -V nem_GumGum_female_chrX_BSQR.g.vcf.gz -V nem_Ngsang_sumatra_female_chrX_BSQR.g.
vcf.gz -V nem_PM1206_haploid_chrX_BSQR.g.vcf.gz -V nem_PM664_haploid_chrX_BSQR.g.vcf.gz -V nem_PM665_haploid_chrX_BSQR.
g.vcf.gz -V nem_Sukai_haploid_chrX_BSQR.g.vcf.gz -V nigra_PF1001_chrX_BSQR.g.vcf.gz -V nigra_PF660_chrX_BSQR.g.vcf.gz -
V nigra_PM1003_haploid_chrX_BSQR.g.vcf.gz -V nigrescens_PM1011_haploid_chrX_BSQR.g.vcf.gz -V nigrescens_PM654_haploid_c
hrX_BSQR.g.vcf.gz -V tog_PF549_chrX_BSQR.g.vcf.gz -V tonk_PF511_chrX_BSQR.g.vcf.gz -V tonk_PF559_chrX_BSQR.g.vcf.gz -V 
tonk_PF563_chrX_BSQR.g.vcf.gz -V tonk_PF597_chrX_BSQR.g.vcf.gz -V tonk_PF626_chrX_BSQR.g.vcf.gz -V tonk_PM592_haploid_c
hrX_BSQR.g.vcf.gz -allsites -o all_diploid_haploid_chrX_BSQR.vcf.gz
```
This then needs to be hard filtered, - first to get rid of the PAR, then for hard filtered parameters, and then using the sliding window apprioach.

First remove PAR:
```
vcftools --gzvcf all_diploid_haploid_chrX_BSQR.vcf.gz --out all_diploid_haploid_chrX_BSQR_noPAR.vcf.gz --chr chrX --from-bp 1691776 --to-bp 148935249 --recode
```

identify indels

```
module load nixpkgs/16.09
module load gatk/3.8


java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T SelectVariants -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/MacaM/MacaM_mt_female.fa -V all_diploid_haploid_chrX_BSQR_noPAR.vcf.gz -selectType INDEL -o all_diploid_haploid_chrX_BSQR_noPAR_indels.vcf.gz

```

now mark the ones to remove (note minimum depth is 5 instead of 10 for chrX)
```
java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T VariantFiltration -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/MacaM/MacaM_mt_female.fa -V all_diploid_haploid_chrX_BSQR_noPAR.vcf.gz --filterExpression "QD < 2.0 || FS > 30.0 || MQ < 40.0 || ReadPosRankSum < -8.0 || ExcessHet > 15.00 || MQRankSum < -10.00 || InbreedingCoeff < 0.00" --filterName "lowqual" -mask all_diploid_haploid_chrX_BSQR_noPAR_indels.vcf.gz --maskExtension 5 --maskName "indel" --genotypeFilterExpression "DP < 5 || DP > 100" --genotypeFilterName "genotypefilter" --setFilteredGtToNocall -o all_diploid_haploid_chrX_BSQR_noPAR_flagged1.vcf.gz
```
now remove them
```
java -jar $EBROOTGATK/GenomeAnalysisTK.jar -T SelectVariants -R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/MacaM/MacaM_mt_female.fa -V all_diploid_haploid_chrX_BSQR_noPAR_flagged1.vcf.gz --excludeFiltered --setFilteredGtToNocall --maxNOCALLfraction 0.5 -o all_diploid_haploid_chrX_BSQR_noPAR_filtered1.vcf.gz
```
and now do sliding window stuff the same as for the autosomes...
