# Hard Filtering

VQSR is not possible with these data, which are limited to 25 genomes.  Two are being sequenced now so I plan to run the pipeline through the existing data, check how the stats look and modify as needed for the final pipeline.  Currently I have post-BQSR genotyping done for everyone for each chr except two males (nem_664 and maura_613) and one female (hecki_644).

Plan is to keep them as chr specific files for all downsteam analyses.  First chr-specific vcfs from each sample need to be combined using CombineGVCFs_bychr, then these all_sample_chr g.vcf files need to be genotyped using GenotypeGVCFs_bychr. A vcf file will be generated with nonvariant reference sites.

Then we can do the hardfiltering.  For this first make an indel file for each chr.  This will be used as a mask plus 5 bp to filter out indels using SelectVariants.  At the same time low quality SNPs will be filtered following GATKs preliminary recommendations.  Then VariantFilteration will generate a filtered vcf file. These files can be inspected to check out statistics.  Then the filtering parameters can be altered if needed.

# Hardfiltering

I am doing a combination of GATK and vcftools for hardfiltering.  Here is the GATK portion:
```
/usr/lib/jvm/java-1.8.0-openjdk.x86_64/bin/java -Djava.io.tmpdir=/scratch/ben/TMP/ -Xmx8g -jar /work/ben/2017_SEAsian_macaques/bin/GenomeAnalysisTK-nightly-2017-10-07-g1994025/GenomeAnalysisTK.jar -T VariantFiltration -R /work/ben/2017_SEAsian_macaques/MacaM/MacaM_mt_y.fa -V /work/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/females_and_males/FandM_chrX_BSQR_jointgeno_allsites_withpapio.vcf.gz --filterExpression "QD < 2.0 || FS > 30.0 || MQ < 40.0 || ReadPosRankSum < -8.0" --filterName "lowqual" -mask /work/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/females_and_males/FandM_chrX_BSQR_jointgeno_allsites_indels_withpapio.vcf.gz --maskExtension 5 --maskName "indel" --genotypeFilterExpression "DP < 10 || DP > 100" --genotypeFilterName "genotypefilter" --setFilteredGtToNocall -o /work/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/females_and_males/FandM_chrX_BSQR_jointgeno_allsites_withpapio_flagged1.vcf.gz
```


# Outputing parameters to inspect distributions

Major insights gleaned from this site: https://evodify.com/gatk-in-non-model-organism/

Output parameters
```
java -Djava.io.tmpdir=/home/ben/scratch/TMP -Xmx2g -jar /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/bin/GenomeAnalysisTK-nightly-2017-10-07-g1994025/GenomeAnalysisTK.jar \
-T VariantsToTable \
-R /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/MacaM/MacaM_mt_y.fa \
-V FandM_chr01_BSQR_jointgeno_SNPsonly_withpapio.vcf.gz \
-F CHROM -F POS -F QUAL -F QD -F DP -F MQ -F MQRankSum -F FS -F ReadPosRankSum -F SOR -F InbreedingCoef -F ExcessHet -F GQ \
-GF DP \
-o GVCFall_SNPs.table
```


Plot distribitions of parameters:
```
setwd('/projects/2017_SEAsian_macaque_genomz/Genotype_Stats')

library('gridExtra')
library('ggplot2')

VCFsnps <- read.csv('GVCFall_SNPs.table', header = T, na.strings=c("","NA"), sep = "\t") 
#VCFindel <- read.csv('GVCFall_INDELs.table', header = T, na.strings=c("","NA"), sep = "\t")
dim(VCFsnps) # this gets the dimensions of the variable above
VCF <- rbind(VCFsnps) # This was used to combine two columns (not needed here)
#VCF <- rbind(VCFsnps, VCFindel)

VCF$Variant <- factor(c(rep("SNPs", dim(VCFsnps)[1]))) # this is a column that has only SNP in it
# modified from this which had two identifer strings in it:
#VCF$Variant <- factor(c(rep("SNPs", dim(VCFsnps)[1]), rep("Indels", dim(VCFindel)[1])))


snps <- '#A9E2E4'

DP <- ggplot(VCF, aes(x=DP, fill=Variant)) + geom_density(alpha=0.3) + 
  geom_vline(xintercept=c(0,300)) + xlim(-3,3000)

ExcessHet <- ggplot(VCF, aes(x=ExcessHet, fill=Variant)) + geom_density(alpha=0.3) + 
  xlim(-3,10)
#range(VCF$ExcessHet)
#quantile(VCF$ExcessHet, probs = c(0.05, 0.99))

InbreedingCoeff <- ggplot(VCF, aes(x=InbreedingCoeff, fill=Variant)) + geom_density(alpha=0.3) + 
  xlim(-3,10)
#range(VCF$InbreedingCoeff,na.rm=TRUE)
#quantile(VCF$InbreedingCoeff, probs = c(0.05, 0.99),na.rm=TRUE)


QD <- ggplot(VCF, aes(x=QD, fill=Variant)) + geom_density(alpha=.3) +
  geom_vline(xintercept=2, size=0.7)

FS <- ggplot(VCF, aes(x=FS, fill=Variant)) + geom_density(alpha=.3) +
  geom_vline(xintercept=c(60, 200), size=0.7) + ylim(0,0.1)

MQ <- ggplot(VCF, aes(x=MQ, fill=Variant)) + geom_density(alpha=.3) +
  geom_vline(xintercept=40, size=0.7)

MQRankSum <- ggplot(VCF, aes(x=MQRankSum, fill=Variant)) + geom_density(alpha=.3) +
  geom_vline(xintercept=-20, size=0.7)

SOR <- ggplot(VCF, aes(x=SOR, fill=Variant)) + geom_density(alpha=.3) +
  geom_vline(xintercept=c(4, 10), size=1, colour = c(snps))

ReadPosRankSum <- ggplot(VCF, aes(x=ReadPosRankSum, fill=Variant)) + geom_density(alpha=.3) +
  geom_vline(xintercept=c(-10,10), size=1, colour = c(snps)) + xlim(-10, 10)

pdf("plot.pdf", height=20, width=15)
theme_set(theme_gray(base_size = 18))
grid.arrange(QD, DP, FS, MQ, MQRankSum, SOR, ReadPosRankSum, ncol=2, nrow =4)
dev.off()
```
