# Hard Filtering

VQSR is not possible with these data, which are limited to 25 genomes.  Two are being sequenced now so I plan to run the pipeline through the existing data, check how the stats look and modify as needed for the final pipeline.  Currently I have post-BQSR genotyping done for everyone for each chr except two males (nem_664 and maura_613) and one female (hecki_644).

Plan is to keep them as chr specific files for all downsteam analyses.  First chr-specific vcfs from each sample need to be combined using CombineGVCFs_bychr, then these all_sample_chr g.vcf files need to be genotyped using GenotypeGVCFs_bychr. A vcf file will be generated with nonvariant reference sites.

Then we can do the hardfiltering.  For this first make an indel file for each chr.  This will be used as a mask plus 5 bp to filter out indels using SelectVariants.  At the same time low quality SNPs will be filtered following GATKs preliminary recommendations.  Then VariantFilteration will generate a filtered vcf file. These files can be inspected to check out statistics.  Then the filtering parameters can be altered if needed.

For now I am still waiting on a few chr from 559 to be genotyped, so I am going to try chr01 for all individuals except the three above.
