# Filtering based on per site coverage

When I analyzed the data using admixfrog, I identified several areas that appeared to be admixed in all species, but these in fact are regions that are duplicated in the nemestrina group but not the rhesus reference genome, so they have higher coverage. higher pi, and higher pairwise divergence (not sure why).  So I am using vcftools to calculate the per site per individual coverage of variable positions.  I wrote a perl script (24_vcftools_descriptivestats_depth_bychr_sqsub_females_and_males.pl) that executes a sbatch script (vcftools_depth_per_site_F_and_M.sh). Both are below:

```perl
#!/usr/bin/perl
# This script will get depth of coverage by site per individual
# Need this before executing:
# export PERL5LIB=/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/bin/vcftools/src/perl/

my $commandline = "module load nixpkgs/16.09";
$status = system($commandline);
$commandline = "module load intel/2018.3";
$status = system($commandline);
$commandline = "module load vcftools/0.1.16";
$status = system($commandline);
$commandline=();

#my $gatkpath = "/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/bin/GenomeAnalysisTK-nightly-2017-10-07-g1994025/";
my $referencegenome="/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/MacaM/MacaM_mt_y.fa";
my $majorpathboth="/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/females_and_males/";

my @chromosomes =("chr01","chr02a","chr02b","chr03","chr04","chr05","chr06","chr07","chr08","chr09","chr10","chr11","chr12","chr
13","chr14","chr15","chr16","chr17","chr18","chr19","chrX");
my @individuals=("bru_PF707","hecki_PF505","hecki_PF643","hecki_PF644","hecki_PF647","hecki_PF648","maura_PF615","maura_PF713","
maura_PM613","maura_PM614","maura_PM616","nem_GumGum_female","nem_Ngsang_sumatra_female","nem_PM1206","nem_PM664","nem_PM665","n
em_Sukai_male","nigra_PF1001","nigra_PF660","nigra_PM1003","nigrescens_PM1011","nigrescens_PM654","tog_PF549","tonk_PF511","tonk
_PF559","tonk_PF563","tonk_PF597","tonk_PF626","tonk_PM592");

foreach my $individual (@individuals){
    foreach my $chromosome (@chromosomes){
	    $commandline="sbatch vcftools_depth_per_site_F_and_M.sh ";
	    $commandline=$commandline.$majorpathboth."FandM_".$chromosome."_BSQR_jointgeno_allsites_filtered_SNPsonly.vcf.gz ";
	    $commandline=$commandline.$majorpathboth."FandM_".$chromosome."_BSQR_jointgeno_allsites_filtered.vcf.gz_".$individua
l."_".$chromosome."_sitedepth ";
	    $commandline=$commandline." ".$individual;
	    print $commandline,"\n\n";
	    $status = system($commandline);
	    $commandline=();
    }
}

```

``` bash
#!/bin/sh
#SBATCH --job-name=vcftools
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=1:00:00
#SBATCH --mem=2gb
#SBATCH --output=vcftools.%J.out
#SBATCH --error=vcftools.%J.err
#SBATCH --account=def-ben

module load nixpkgs/16.09
module load intel/2018.3
module load vcftools/0.1.16

vcftools --gzvcf ${1} --out ${2} --indv ${3} --site-depth
```
