# dNdS Analysis

One reviewer suggested to do a PAML analysis on the Ninteract genes and a subset of the autosomal genes.  I think this is a great idea and I plan to do this on all genes.

One idea I had was to use the mtDNA phylogeny in this analysis. One could also use the species tree, but there isn't really strong support for an individual species tree.  Another possibility is to use individual trees for each gene, but this also is problematic because many genes won't have a strongly supported tree.  A fourth possibility is to use a 'star' tree. I like the first alternative most

Another question is how to deal with heterozygous positions.  I think I will develop a script that inserts IUPAC characters for these (which PAML will accept I think - need to confirm).

The first step is to get the coordinates of one transcript from all genes, and keep track of the direction of transcription in the file name.  I did this using this perl script:
```
#!/usr/bin/env perl
use strict;
use warnings;


#  This program reads in gff file and outputs
#  the coordinates of all CDS for each gene, but only for "transcript_01"
#  Other transcripts of each gene are not included; this ensures that each gene is analyzed only once

# to execute type ./Get_coordinates_of_CDS_in_each_gene_outout_individual_beds.pl MacaM_Rhesus_Genome_Annotation_v7.6.8.gff


my $inputfile = $ARGV[0];
my $outputfile;

unless (open DATAINPUT, $inputfile) {
	print "Can not find the input file.\n";
	exit;
}

my @temp;
my @temp1;
my @Gene_name;
my @Gene_ID;
my $Gene_name;
my $Gene_ID;
my $Transcript_ID;
my %gene_hash;
my $gene_counter=0;
my $exon_counter;
while ( my $line = <DATAINPUT>) {
	chomp($line);
	@temp=split('\t',$line);
	#print $temp[1],"\n";
	if($temp[2] eq 'mRNA'){ # we need to parse the gene name and geneID
		# We are assuming that all CDs are preceded by a "mRNA" annotation
		@temp1=split(';',$temp[8]);
		@Gene_name=split('=',$temp1[0]);
		$Gene_name=$Gene_name[1];
		@Gene_ID=split('GID=',$temp1[1]);
		$Gene_ID=$Gene_ID[1];
		print $Gene_name,"\n";
		$gene_counter+=1;
		$exon_counter=1;
	}
	elsif($temp[2] eq 'CDS'){ # we need to save the coordinates and phase of this CDS
		# first get the transcript id
		@temp1=split(';',$temp[8]);
		@Gene_name=split('=',$temp1[0]);
		$Gene_name=$Gene_name[1];
		@Gene_ID=split('GID=',$temp1[1]);
		$Gene_ID=$Gene_ID[1];
		print $Gene_name,"\n";
		# only consider the first transcript
		if($Gene_name =~ /transcript_01/){
			$Transcript_ID="transcript_01";
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[0]=$temp[3]; # start coordinate
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[1]=$temp[4];	# stop coordinate
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[2]=$temp[7];	# phase sort of
									#'0' indicates that the first base of the feature is the first base of a codon, 
									#'1' that the second base is the first base of a codon,
									#'2' that the third base is the first base of a codon,
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[4]=$temp[0]; # chromosome						
			if($temp[6] eq '+'){
				$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[3]=1;	# forward orientation

			}
			elsif($temp[6] eq '-'){
				$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[3]=-1;	# reverse orientation
			}
			else{
				print "WTF ",$temp[6],"\n";
			}
			$exon_counter+=1;
		}
	}
} # end while	
close DATAINPUT;	
# OK, now all the CDS are in a hash
my $switch=0;
# print a bed file for each transcript (sometimes there will be multiple transcripts for the same gene)
foreach my $key (sort keys %gene_hash){
	foreach my $transcript_id (sort keys %{$gene_hash{$key}}){
		foreach my $exon (sort keys %{$gene_hash{$key}{$transcript_id}}){
			if($switch == 0){ # open a new file for this transcript
				$switch=1;
				if($gene_hash{$key}{$transcript_id}{$exon}[3] eq 1){ # assume all exons are in this orientation
					unless (open(OUTFILE, ">".$key."_".$transcript_id.".coord"))  {
						print "I can\'t write to $outputfile\n";
						exit;
					}
					print "Creating output file: ".$key."_".$transcript_id.".coord\n";
				}
				else{ # assume all exons are in this orientation
					unless (open(OUTFILE, ">".$key."_".$transcript_id."_rc.coord"))  {
						print "I can\'t write to $outputfile\n";
						exit;
					}
					print "Creating output file: ".$key."_".$transcript_id."_rc.coord\n";
				}	
			}	
			print OUTFILE $gene_hash{$key}{$transcript_id}{$exon}[4],"\t";
			print OUTFILE $gene_hash{$key}{$transcript_id}{$exon}[0],"\t";
			print OUTFILE $gene_hash{$key}{$transcript_id}{$exon}[1],"\n";
		}
		close OUTFILE;
		$switch=0; # reset for next transcript
	}		
}	

```
The coordinates from each gene (n = 16049) are on graham here:
```
/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/MacaM/CDS_coordinates_individual_genes
```

# Generate vcf files

First I had to put each bed file into its own directory according to which chr the gene is on.  That way I could run the script to make a tab file for each gene for each chromosome.
```
mkdir chr01...
mkdir chrX
```
OK now I have a directory for each chr.  Now search for files that have each chr and move them to the appropriate directory if they do.
```
grep -lir 'chrX' ./* | xargs mv -t chrX
```
Now run bcftools on directories of bed files using this perl scripe (`Run_bcftools_with_lots_of_inputs.pl`)"
```
#!/usr/bin/env perl
use strict;
use warnings;


#  This program reads in coordinate files from a directory
# and feeds them into a bash script that extracts sections using bcftools

# to execute type ./Run_bcftools_with_lots_of_inputs.pl path_to_bed_files chr


my $inputfile = $ARGV[0];
	unless (open DATAINPUT, $inputfile) {
		print "Can not find the input file.\n";
		exit;
	}
my $chr = $ARGV[1];
my @files = glob($inputfile.'/*coord');

foreach ( @files ) {
	print $_,"\n";
	system( "./2021_bcftools_extract_sections_from_vcf.sh $_ $chr")
}
```

where this bash script (`2021_bcftools_extract_sections_from_vcf.sh`) looks like this:
```
#!/bin/sh
#SBATCH --job-name=bcftools
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=0:10:00
#SBATCH --mem=2gb
#SBATCH --output=bcftools.%J.out
#SBATCH --error=bcftools.%J.err
#SBATCH --account=def-ben

# execute like this: ./2021_bcftools_extract_sections_from_vcf.sh path_and_filename_of_coordinate_file chr

module load StdEnv/2020 gcc/9.3.0 bcftools/1.11
bcftools view -R ${1} /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio/FandM_$2_BSQR
_jointgeno_allsites_withpapio.vcf.gz -o ${1}.vcf
```

# Generate tab files

# Generate paml files from the tab files

# Do PAML analysis

# Compare total branch length of Ninteract genes to non-Ninteract genes
