# dNdS Analysis

The first step is to get the coordinates of one transcript from all genes, and keep track of the direction of transcription in the file name.  I did this using this perl script (Get_coordinates_of_CDS_in_each_gene_outout_individual_beds.pl):
```
#!/usr/bin/env perl
use strict;
use warnings;


#  This program reads in gff file and outputs
#  the coordinates of all CDS for each gene, but only for "transcript_01"
#  Other transcripts of each gene are not included; this ensures that each gene is analyzed only once

# to execute type ./Get_coordinates_of_CDS_in_each_gene_outout_individual_beds.pl MacaM_Rhesus_Genome_Annotation_v7.6.8.gff


my $inputfile = $ARGV[0];
my $outputfile;

unless (open DATAINPUT, $inputfile) {
	print "Can not find the input file.\n";
	exit;
}

my @temp;
my @temp1;
my @Gene_name;
my @Gene_ID;
my $Gene_name;
my $Gene_ID;
my $Transcript_ID;
my %gene_hash;
my $gene_counter=0;
my $exon_counter;
while ( my $line = <DATAINPUT>) {
	chomp($line);
	@temp=split('\t',$line);
	#print $temp[1],"\n";
	if($temp[2] eq 'mRNA'){ # we need to parse the gene name and geneID
		# We are assuming that all CDs are preceded by a "mRNA" annotation
		@temp1=split(';',$temp[8]);
		@Gene_name=split('=',$temp1[0]);
		$Gene_name=$Gene_name[1];
		@Gene_ID=split('GID=',$temp1[1]);
		$Gene_ID=$Gene_ID[1];
		print $Gene_name,"\n";
		$gene_counter+=1;
		$exon_counter=1;
	}
	elsif($temp[2] eq 'CDS'){ # we need to save the coordinates and phase of this CDS
		# first get the transcript id
		@temp1=split(';',$temp[8]);
		@Gene_name=split('=',$temp1[0]);
		$Gene_name=$Gene_name[1];
		@Gene_ID=split('GID=',$temp1[1]);
		$Gene_ID=$Gene_ID[1];
		print $Gene_name,"\n";
		# only consider the first transcript
		if($Gene_name =~ /transcript_01/){
			$Transcript_ID="transcript_01";
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[0]=$temp[3]; # start coordinate
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[1]=$temp[4];	# stop coordinate
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[2]=$temp[7];	# phase sort of
									#'0' indicates that the first base of the feature is the first base of a codon, 
									#'1' that the second base is the first base of a codon,
									#'2' that the third base is the first base of a codon,
			$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[4]=$temp[0]; # chromosome						
			if($temp[6] eq '+'){
				$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[3]=1;	# forward orientation

			}
			elsif($temp[6] eq '-'){
				$gene_hash{$Gene_name."_".$Gene_ID}{$Transcript_ID}{$exon_counter}[3]=-1;	# reverse orientation
			}
			else{
				print "WTF ",$temp[6],"\n";
			}
			$exon_counter+=1;
		}
	}
} # end while	
close DATAINPUT;	
# OK, now all the CDS are in a hash
my $switch=0;
# print a bed file for each transcript (sometimes there will be multiple transcripts for the same gene)
foreach my $key (sort keys %gene_hash){
	foreach my $transcript_id (sort keys %{$gene_hash{$key}}){
		foreach my $exon (sort {$a <=> $b} keys %{$gene_hash{$key}{$transcript_id}}){
			if($switch == 0){ # open a new file for this transcript
				$switch=1;
				if($gene_hash{$key}{$transcript_id}{$exon}[3] eq 1){ # assume all exons are in this orientation
					unless (open(OUTFILE, ">".$key."_".$transcript_id.".coord"))  {
						print "I can\'t write to $outputfile\n";
						exit;
					}
					print "Creating output file: ".$key."_".$transcript_id.".coord\n";
				}
				else{ # assume all exons are in this orientation
					unless (open(OUTFILE, ">".$key."_".$transcript_id."_rc.coord"))  {
						print "I can\'t write to $outputfile\n";
						exit;
					}
					print "Creating output file: ".$key."_".$transcript_id."_rc.coord\n";
				}	
			}	
			print OUTFILE $gene_hash{$key}{$transcript_id}{$exon}[4],"\t";
			print OUTFILE $gene_hash{$key}{$transcript_id}{$exon}[0],"\t";
			print OUTFILE $gene_hash{$key}{$transcript_id}{$exon}[1],"\n";
		}
		close OUTFILE;
		$switch=0; # reset for next transcript
	}		
}	

```
The coordinates from each gene (n = 16049) are on graham here:
```
/home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/MacaM/CDS_coordinates_individual_genes
```

# Generate vcf files

First I had to put each bed file into its own directory according to which chr the gene is on.  That way I could run the script to make a tab file for each gene for each chromosome.
```
mkdir chr01...
mkdir chrX
```
Now search for files that have each chr and move them to the appropriate directory if they do. For example, for chrX:
```
grep -lir 'chrX' ./* | xargs mv -t chrX
```
Now run bcftools on directories of bed files using this perl script (`Run_bcftools_with_lots_of_inputs.pl`)"
```
#!/usr/bin/env perl
use strict;
use warnings;


#  This program reads in coordinate files from a directory
# and feeds them into a bash script that extracts sections using bcftools

# to execute type ./Run_bcftools_with_lots_of_inputs.pl path_to_bed_files chr


my $inputfile = $ARGV[0];
	unless (open DATAINPUT, $inputfile) {
		print "Can not find the input file.\n";
		exit;
	}
my $chr = $ARGV[1];
my @files = glob($inputfile.'/*coord');

foreach ( @files ) {
	print $_,"\n";
	system( "./2021_bcftools_extract_sections_from_vcf.sh $_ $chr")
}
```

where this bash script (`2021_bcftools_extract_sections_from_vcf.sh`) looks like this:
```
#!/bin/sh
#SBATCH --job-name=bcftools
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=0:10:00
#SBATCH --mem=2gb
#SBATCH --output=bcftools.%J.out
#SBATCH --error=bcftools.%J.err
#SBATCH --account=def-ben

# execute like this: ./2021_bcftools_extract_sections_from_vcf.sh path_and_filename_of_coordinate_file chr
# load these modules before running:
# module load StdEnv/2020 gcc/9.3.0 bcftools/1.11
bcftools view -R ${1} /home/ben/projects/rrg-ben/ben/2017_SEAsian_macaques/SEAsian_macaques_bam/with_papio/FandM_$2_BSQR_jointgeno_allsites_withpapio.vcf.gz -o ${1}.vcf
```

# Compress coord directory to save space
Compress a directory like this:
```
for dir in FILENAME; do base=$(basename "$dir"); tar -czf "${base}.tar.gz" --remove-files "$dir"; done
```
You can view files in the archive like this:
```
tar -tf coords.tar.gz
```
You can view individual files in the compressed archive like this:
```
tar -Oxvf coords.tar.gz coords/ZC3H11A_transcript_01_9877_transcript_01_rc.coord | less
```

# Generate tab files
THis is done with a perl script (vcf2tab.pl):
```
#!/usr/bin/env perl
use strict;
use warnings;


# This program makes tab files from each vcf file in a directory

# before executing load modules
# module load nixpkgs/16.09 intel/2018.3 vcftools/0.1.16
# module load StdEnv/2020 perl/5.30.2
# to execute type ./vcf2tab.pl path_to_vcf_files


    
my $inputfile = $ARGV[0];
	unless (open DATAINPUT, $inputfile) {
		print "Can not find the input file.\n";
		exit;
	}

my @files = glob($inputfile.'/*vcf');

foreach ( @files ) {
 #   print $_,"\n";
	system( "vcf-to-tab < $_ > $_.tab")
}
```

# Generate paml files from the tab files

This script runs another perl script on each tab file (Make_lots_of_paml_files.pl):
```
#!/usr/bin/env perl
use strict;
use warnings;


#  This program reads in coordinate files from a directory
# and feeds them into a bash script that extracts sections using bcftools

# before executing load modules
# module load StdEnv/2020 perl/5.30.2
# to execute type ./Make_lots_of_paml_files.pl path_to_tab_files


    
my $inputfile = $ARGV[0];
	unless (open DATAINPUT, $inputfile) {
		print "Can not find the input file.\n";
		exit;
	}

my @files = glob($inputfile.'/*tab');

foreach ( @files ) {
 #   print $_,"\n";
	system( "./tab2paml.pl $_ $_\.paml_in")
}	
```
and heres the `tab2paml.pl` script:
```
#!/usr/bin/env perl
use strict;
use warnings;
use List::MoreUtils qw/ uniq /;
use List::Util qw( min max );


#  This program reads in a tab delimited genotype file generated
#  from vcftools and prints out a paml file 

# Data should be from one chromosome.

# on graham, you need to load perl first:
# module load StdEnv/2020 perl/5.30.2

# to execute type tab2paml.pl inputfile.tab output_paml_in 

# From the paml doc:
# In a sequence, T, C, A, G, U, t, c, a, g, u are recognized as nucleotides (for baseml, basemlg and
# codonml), while the standard one-letter codes (A, R, N, D, C, Q, E, G, H, I, L, K, M, F, P, S, T, W,
# Y, V or their lowercase equivalents) are recognized as amino acids. Ambiguity characters
# (undetermined nucleotides or amino acids) are allowed as well. Three special characters ".", "-", and
# "?" are interpreted like this: a dot means the same character as in the first sequence, a dash means an
# alignment gap, and a question mark means an undetermined nucleotide or amino acid.

# if cleandata = 0, both ambiguity characters and alignment gaps are treated
# as ambiguity characters. In the pairwise distance calculation (the lower-diagonal distance matrix in
# the output), cleandata = 1 means “complete deletion”, with all sites involving ambiguity characters
# and alignment gaps removed from all sequences, while cleandata = 0 means “pairwise deletion”,
# with only sites which have missing characters in the pair removed.

# Thus this script will generate one seq for each diploid sequence and use ambiguity characters for heterozygous
# positions, including Ns for deletions or heterozygous deletions

# The analysis will use "cleandata = 0"

# the analysis will not include the baboon seq.


my $inputfile = $ARGV[0];
my $outputfile2 = $ARGV[1];

unless (open DATAINPUT, $inputfile) {
	print "Can not find the input file.\n";
	exit;
}


unless (open(OUTFILE2, ">$outputfile2"))  {
	print "I can\'t write to $outputfile2\n";
	exit;
}
print "Creating output file: $outputfile2\n";


my $bru_PF707="";
my $download="";
my $hecki_PF505="";
my $hecki_PF643="";
my $hecki_PF644="";
my $hecki_PF647="";
my $hecki_PF648="";
my $maura_PF615="";
my $maura_PF713="";
my $maura_PM613="";
my $maura_PM614="";
my $maura_PM616="";
my $nem_GumGum_female="";
my $nem_Ngsang_sumatra_female="";
my $nem_PM1206="";
my $nem_PM664="";
my $nem_PM665="";
my $nem_Sukai_male="";
my $nigra_PF1001="";
my $nigra_PF660="";
my $nigra_PM1003="";
my $nigrescens_PM1011="";
my $nigrescens_PM654="";
my $tog_PF549="";
my $tonk_PF511="";
my $tonk_PF559="";
my $tonk_PF563="";
my $tonk_PF597="";
my $tonk_PF626="";
my $tonk_PM592="";

my @alleles;
my @temp;
my @lengths;
my $max;
my $a;
my $b;
my $start;

while ( my $line = <DATAINPUT>) {
	chomp($line);
	@temp=split(/[\/'\t']+/,$line);
	if($temp[0] ne '#CHROM'){
		# print $temp[8],"\n";
		# load the alleles
			# load each of the 60 alleles
			for($a=3; $a<=62; $a++){
				$alleles[$a-3]=$temp[$a]; # start index at zero
			}	
			# now see which is the longest
			@lengths=();
			for($a=0; $a<60; $a++){
				push(@lengths,length($alleles[$a]));
			}
			$max = max @lengths;
			#print $max;
			# pad alleles if needed
			my @temp_alleles;
			for($a=0; $a<60; $a++){
				@temp_alleles=split(//,$alleles[$a]);
				if(($#temp_alleles+1) < $max){
					for($b=($#temp_alleles+1); $b<$max; $b++){
						$alleles[$a]=$alleles[$a]."-";
					}	
				}	
			}		
			# now the lengths should all be the same

			# add the data to the chrs
			if($alleles[0] eq $alleles[1]){
				$bru_PF707=$bru_PF707.$alleles[0];
			}
			else{
				for($b=0; $b<$max; $b++){
					$bru_PF707=$bru_PF707."N";
				}
			}	
			if($alleles[2] eq $alleles[3]){
				$download=$download.$alleles[2];
			}
			else{
				for($b=0; $b<$max; $b++){
					$download=$download."N";
				}
			}	
			if($alleles[4] eq $alleles[5]){
				$hecki_PF505=$hecki_PF505.$alleles[4];
			}
			else{
				for($b=0; $b<$max; $b++){
					$hecki_PF505=$hecki_PF505."N";
				}
			}	
			if($alleles[6] eq $alleles[7]){
				$hecki_PF643=$hecki_PF643.$alleles[6];
			}
			else{
				for($b=0; $b<$max; $b++){
					$hecki_PF643=$hecki_PF643."N";
				}
			}	
			if($alleles[8] eq $alleles[9]){
				$hecki_PF644=$hecki_PF644.$alleles[8];
			}
			else{
				for($b=0; $b<$max; $b++){
					$hecki_PF644=$hecki_PF644."N";
				}
			}	
			if($alleles[10] eq $alleles[11]){
				$hecki_PF647=$hecki_PF647.$alleles[10];
			}
			else{
				for($b=0; $b<$max; $b++){
					$hecki_PF647=$hecki_PF647."N";
				}
			}	
			if($alleles[12] eq $alleles[13]){
				$hecki_PF648=$hecki_PF648.$alleles[12];
			}
			else{
				for($b=0; $b<$max; $b++){
					$hecki_PF648=$hecki_PF648."N";
				}
			}	
			if($alleles[14] eq $alleles[15]){
				$maura_PF615=$maura_PF615.$alleles[14];
			}
			else{
				for($b=0; $b<$max; $b++){
					$maura_PF615=$maura_PF615."N";
				}
			}	
			if($alleles[16] eq $alleles[17]){
				$maura_PF713=$maura_PF713.$alleles[16];
			}
			else{
				for($b=0; $b<$max; $b++){
					$maura_PF713=$maura_PF713."N";
				}
			}	
			if($alleles[18] eq $alleles[19]){
				$maura_PM613=$maura_PM613.$alleles[18];
			}
			else{
				for($b=0; $b<$max; $b++){
					$maura_PM613=$maura_PM613."N";
				}
			}
			if($alleles[20] eq $alleles[21]){
				$maura_PM614=$maura_PM614.$alleles[20];
			}
			else{
				for($b=0; $b<$max; $b++){
					$maura_PM614=$maura_PM614."N";
				}
			}
			if($alleles[22] eq $alleles[23]){
				$maura_PM616=$maura_PM616.$alleles[22];
			}
			else{
				for($b=0; $b<$max; $b++){
					$maura_PM616=$maura_PM616."N";
				}
			}	
			if($alleles[24] eq $alleles[25]){
				$nem_GumGum_female=$nem_GumGum_female.$alleles[24];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nem_GumGum_female=$nem_GumGum_female."N";
				}
			}	
			if($alleles[26] eq $alleles[27]){
				$nem_Ngsang_sumatra_female=$nem_Ngsang_sumatra_female.$alleles[26];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nem_Ngsang_sumatra_female=$nem_Ngsang_sumatra_female."N";
				}
			}	
			if($alleles[28] eq $alleles[29]){
				$nem_PM1206=$nem_PM1206.$alleles[28];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nem_PM1206=$nem_PM1206."N";
				}
			}	
			if($alleles[30] eq $alleles[31]){
				$nem_PM664=$nem_PM664.$alleles[30];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nem_PM664=$nem_PM664."N";
				}
			}	
			if($alleles[32] eq $alleles[33]){
				$nem_PM665=$nem_PM665.$alleles[32];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nem_PM665=$nem_PM665."N";
				}
			}	
			if($alleles[34] eq $alleles[35]){
				$nem_Sukai_male=$nem_Sukai_male.$alleles[34];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nem_Sukai_male=$nem_Sukai_male."N";
				}
			}
			if($alleles[36] eq $alleles[37]){
				$nigra_PF1001=$nigra_PF1001.$alleles[36];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nigra_PF1001=$nigra_PF1001."N";
				}
			}	
			if($alleles[38] eq $alleles[39]){
				$nigra_PF660=$nigra_PF660.$alleles[38];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nigra_PF660=$nigra_PF660."N";
				}
			}	
			if($alleles[40] eq $alleles[41]){
				$nigra_PM1003=$nigra_PM1003.$alleles[40];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nigra_PM1003=$nigra_PM1003."N";
				}
			}	
			if($alleles[42] eq $alleles[43]){
				$nigrescens_PM1011=$nigrescens_PM1011.$alleles[42];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nigrescens_PM1011=$nigrescens_PM1011."N";
				}
			}	
			if($alleles[44] eq $alleles[45]){
				$nigrescens_PM654=$nigrescens_PM654.$alleles[44];
			}
			else{
				for($b=0; $b<$max; $b++){
					$nigrescens_PM654=$nigrescens_PM654."N";
				}
			}	
			if($alleles[46] eq $alleles[47]){
				$tog_PF549=$tog_PF549.$alleles[46];
			}
			else{
				for($b=0; $b<$max; $b++){
					$tog_PF549=$tog_PF549."N";
				}
			}	
			if($alleles[48] eq $alleles[49]){
				$tonk_PF511=$tonk_PF511.$alleles[48];
			}
			else{
				for($b=0; $b<$max; $b++){
					$tonk_PF511=$tonk_PF511."N";
				}
			}	
			if($alleles[50] eq $alleles[51]){
				$tonk_PF559=$tonk_PF559.$alleles[50];
			}
			else{
				for($b=0; $b<$max; $b++){
					$tonk_PF559=$tonk_PF559."N";
				}
			}	
			if($alleles[52] eq $alleles[53]){
				$tonk_PF563=$tonk_PF563.$alleles[52];
			}
			else{
				for($b=0; $b<$max; $b++){
					$tonk_PF563=$tonk_PF563."N";
				}
			}	
			if($alleles[54] eq $alleles[55]){
				$tonk_PF597=$tonk_PF597.$alleles[54];
			}
			else{
				for($b=0; $b<$max; $b++){
					$tonk_PF597=$tonk_PF597."N";
				}
			}	
			if($alleles[56] eq $alleles[57]){
				$tonk_PF626=$tonk_PF626.$alleles[56];
			}
			else{
				for($b=0; $b<$max; $b++){
					$tonk_PF626=$tonk_PF626."N";
				}
			}
			if($alleles[58] eq $alleles[59]){
				$tonk_PM592=$tonk_PM592.$alleles[58];
			}
			else{
				for($b=0; $b<$max; $b++){
					$tonk_PM592=$tonk_PM592."N";
				}
			}	
			# should be done adding bases to each seq
	} # end if to check for header of input file
} # end while	
close DATAINPUT;				

# print out the lengths of each chr to see if we missed anything
# print "Length of bru707 :",length($bru_PF707),"\n";
# print "Length of tonk_PM592 :",length($tonk_PM592),"\n";
# print "Length of tonk_PF511 :",length($tonk_PF511),"\n";

# reversecomplement if needed
if(substr($inputfile,-17) eq "_rc.coord.vcf.tab"){
	my $bru_PF707_revcom=reverse $bru_PF707;
	my $download_revcom=reverse $download;
	my $hecki_PF505_revcom=reverse $hecki_PF505;
	my $hecki_PF643_revcom=reverse $hecki_PF643;
	my $hecki_PF644_revcom=reverse $hecki_PF644;
	my $hecki_PF647_revcom=reverse $hecki_PF647;
	my $hecki_PF648_revcom=reverse $hecki_PF648;
	my $maura_PF615_revcom=reverse $maura_PF615;
	my $maura_PF713_revcom=reverse $maura_PF713;
	my $maura_PM613_revcom=reverse $maura_PM613;
	my $maura_PM614_revcom=reverse $maura_PM614;
	my $maura_PM616_revcom=reverse $maura_PM616;
	my $nem_GumGum_female_revcom=reverse $nem_GumGum_female;
	my $nem_Ngsang_sumatra_female_revcom=reverse $nem_Ngsang_sumatra_female;
	my $nem_PM1206_revcom=reverse $nem_PM1206;
	my $nem_PM664_revcom=reverse $nem_PM664;
	my $nem_PM665_revcom=reverse $nem_PM665;
	my $nem_Sukai_male_revcom=reverse $nem_Sukai_male;
	my $nigra_PF1001_revcom=reverse $nigra_PF1001;
	my $nigra_PF660_revcom=reverse $nigra_PF660;
	my $nigra_PM1003_revcom=reverse $nigra_PM1003;
	my $nigrescens_PM1011_revcom=reverse $nigrescens_PM1011;
	my $nigrescens_PM654_revcom=reverse $nigrescens_PM654;
	my $tog_PF549_revcom=reverse $tog_PF549;
	my $tonk_PF511_revcom=reverse $tonk_PF511;
	my $tonk_PF559_revcom=reverse $tonk_PF559;
	my $tonk_PF563_revcom=reverse $tonk_PF563;
	my $tonk_PF597_revcom=reverse $tonk_PF597;
	my $tonk_PF626_revcom=reverse $tonk_PF626;
	my $tonk_PM592_revcom=reverse $tonk_PM592;
	$bru_PF707_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$download_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$hecki_PF505_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$hecki_PF643_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$hecki_PF644_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$hecki_PF647_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$hecki_PF648_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$maura_PF615_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$maura_PF713_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$maura_PM613_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$maura_PM614_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$maura_PM616_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nem_GumGum_female_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nem_Ngsang_sumatra_female_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nem_PM1206_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nem_PM664_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nem_PM665_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nem_Sukai_male_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nigra_PF1001_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nigra_PF660_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nigra_PM1003_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nigrescens_PM1011_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$nigrescens_PM654_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$tog_PF549_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$tonk_PF511_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$tonk_PF559_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$tonk_PF563_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$tonk_PF597_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$tonk_PF626_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$tonk_PM592_revcom =~ tr/ACGTacgt/TGCAtgca/;
	$bru_PF707 = $bru_PF707_revcom;
	$download = $download_revcom;
	$hecki_PF505 = $hecki_PF505_revcom;
	$hecki_PF643 = $hecki_PF643_revcom;
	$hecki_PF644 = $hecki_PF644_revcom;
	$hecki_PF647 = $hecki_PF647_revcom;
	$hecki_PF648 = $hecki_PF648_revcom;
	$maura_PF615 = $maura_PF615_revcom;
	$maura_PF713 = $maura_PF713_revcom;
	$maura_PM613 = $maura_PM613_revcom;
	$maura_PM614 = $maura_PM614_revcom;
	$maura_PM616 = $maura_PM616_revcom;
	$nem_GumGum_female = $nem_GumGum_female_revcom;
	$nem_Ngsang_sumatra_female = $nem_Ngsang_sumatra_female_revcom;
	$nem_PM1206 = $nem_PM1206_revcom;
	$nem_PM664 = $nem_PM664_revcom;
	$nem_PM665 = $nem_PM665_revcom;
	$nem_Sukai_male = $nem_Sukai_male_revcom;
	$nigra_PF1001 = $nigra_PF1001_revcom;
	$nigra_PF660 = $nigra_PF660_revcom;
	$nigra_PM1003 = $nigra_PM1003_revcom;
	$nigrescens_PM1011 = $nigrescens_PM1011_revcom;
	$nigrescens_PM654 = $nigrescens_PM654_revcom;
	$tog_PF549 = $tog_PF549_revcom;
	$tonk_PF511 = $tonk_PF511_revcom;
	$tonk_PF559 = $tonk_PF559_revcom;
	$tonk_PF563 = $tonk_PF563_revcom;
	$tonk_PF597 = $tonk_PF597_revcom;
	$tonk_PF626 = $tonk_PF626_revcom;
	$tonk_PM592 = $tonk_PM592_revcom;
}


# OK print out the fasta file

print OUTFILE2 "29 ",length($bru_PF707),"\n";
print OUTFILE2 "bru_PF707     ";
print OUTFILE2 $bru_PF707,"\n";
print OUTFILE2 "hecki_PF505     ";
print OUTFILE2 $hecki_PF505,"\n";
print OUTFILE2 "hecki_PF643     ";
print OUTFILE2 $hecki_PF643,"\n";
print OUTFILE2 "hecki_PF644     ";
print OUTFILE2 $hecki_PF644,"\n";
print OUTFILE2 "hecki_PF647     ";
print OUTFILE2 $hecki_PF647,"\n";
print OUTFILE2 "hecki_PF648     ";
print OUTFILE2 $hecki_PF648,"\n";
print OUTFILE2 "maura_PF615     ";
print OUTFILE2 $maura_PF615,"\n";
print OUTFILE2 "maura_PF713     ";
print OUTFILE2 $maura_PF713,"\n";
print OUTFILE2 "maura_PM613     ";
print OUTFILE2 $maura_PM613,"\n";
print OUTFILE2 "maura_PM614     ";
print OUTFILE2 $maura_PM614,"\n";
print OUTFILE2 "maura_PM616     ";
print OUTFILE2 $maura_PM616,"\n";
print OUTFILE2 "nem_GumGum_female     ";
print OUTFILE2 $nem_GumGum_female,"\n";
print OUTFILE2 "nem_Ngsang_sumatra_female     ";
print OUTFILE2 $nem_Ngsang_sumatra_female,"\n";
print OUTFILE2 "nem_PM1206     ";
print OUTFILE2 $nem_PM1206,"\n";
print OUTFILE2 "nem_PM664     ";
print OUTFILE2 $nem_PM664,"\n";
print OUTFILE2 "nem_PM665     ";
print OUTFILE2 $nem_PM665,"\n";
print OUTFILE2 "nem_Sukai_male     ";
print OUTFILE2 $nem_Sukai_male,"\n";
print OUTFILE2 "nigra_PF1001     ";
print OUTFILE2 $nigra_PF1001,"\n";
print OUTFILE2 "nigra_PF660     ";
print OUTFILE2 $nigra_PF660,"\n";
print OUTFILE2 "nigra_PM1003     ";
print OUTFILE2 $nigra_PM1003,"\n";
print OUTFILE2 "nigrescens_PM1011     ";
print OUTFILE2 $nigrescens_PM1011,"\n";
print OUTFILE2 "nigrescens_PM654     ";
print OUTFILE2 $nigrescens_PM654,"\n";
print OUTFILE2 "tog_PF549     ";
print OUTFILE2 $tog_PF549,"\n";
print OUTFILE2 "tonk_PF511     ";
print OUTFILE2 $tonk_PF511,"\n";
print OUTFILE2 "tonk_PF559     ";
print OUTFILE2 $tonk_PF559,"\n";
print OUTFILE2 "tonk_PF563     ";
print OUTFILE2 $tonk_PF563,"\n";
print OUTFILE2 "tonk_PF597     ";
print OUTFILE2 $tonk_PF597,"\n";
print OUTFILE2 "tonk_PF626     ";
print OUTFILE2 $tonk_PF626,"\n";
print OUTFILE2 "tonk_PM592     ";
print OUTFILE2 $tonk_PM592,"\n";



close OUTFILE2;
my $n = length($bru_PF707)/3;

if( $n != int(length($bru_PF707)/3) ){
	print "This infile not multiples of 3: ",$inputfile," ",length($bru_PF707),"\n";
}
```
# Do PAML analysis
Use an sbatch file:
```
#!/bin/sh
#SBATCH --job-name=pamlcommando
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --time=168:00:00
#SBATCH --mem=2gb
#SBATCH --output=pamlcommando.%J.out
#SBATCH --error=pamlcommando.%J.err
#SBATCH --account=def-ben

# execute like this: sbatch paml_commando.sh path_and_filename_of_directory_with_paml_files
# e.g., sbatch paml_commando.sh ../CDS_coordinates_individual_genes/chr01/tabs
# load these modules before running:
module load StdEnv/2020 perl/5.30.2
module load StdEnv/2020  gcc/9.3.0 paml/4.9j

./Execute_lots_of_paml_runs.pl ${1}
```

# Compare total branch length of Ninteract genes to non-Ninteract genes
